web/interceptor/tokens.go

```diff
@@ -0,0 +1,75 @@
+package interceptor
+
+import (
+	"context"
+	"log"
+	"strings"
+
+	"github.com/quickfeed/quickfeed/web/auth"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+type (
+	// The userIDs interface must be implemented by request types that may need to update the tokens.
+	userIDs interface{ UserIDs() []uint64 }
+	// Marker interface to detect the GroupRequest type needed for DeleteGroup.
+	isGroup interface{ GetGroupID() uint64 }
+)
+
+var defaultTokenUpdater = func(_ context.Context, tm *auth.TokenManager, msg userIDs) error {
+	for _, userID := range msg.UserIDs() {
+		if err := tm.Add(userID); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// tokenUpdateMethods is a map of methods that require updating the list of users who need a new JWT.
+var tokenUpdateMethods = map[string]func(context.Context, *auth.TokenManager, userIDs) error{
+	"UpdateUser":        defaultTokenUpdater, // User has been promoted to admin or demoted.
+	"UpdateGroup":       defaultTokenUpdater, // Users added to a group or removed from a group.
+	"UpdateEnrollments": defaultTokenUpdater, // User enrolled into a new course or promoted to TA.
+
+	"CreateCourse": // The signed in user gets the teacher role in the new course.
+	func(ctx context.Context, tm *auth.TokenManager, _ userIDs) error {
+		claims, err := tm.GetClaims(ctx)
+		if err != nil {
+			return err
+		}
+		return tm.Add(claims.UserID)
+	},
+
+	"DeleteGroup": // Group members removed from the group.
+	func(ctx context.Context, tm *auth.TokenManager, msg userIDs) error {
+		if grp, ok := msg.(isGroup); ok {
+			group, err := tm.Database().GetGroup(grp.GetGroupID())
+			if err != nil {
+				return err
+			}
+			return defaultTokenUpdater(ctx, tm, group)
+		}
+		return nil // TODO(meling) This should probably be an error??
```

Comment:
Should this return `nil` or an error? With the current tests it is unreachable. That may be fine... but it is not clear to me what should happen if the `isGroup` interface is not implemented.... Seems like a programmer error that we should return an error for...

Reply:
Agree, returning an error now.